Feat: sway-mode module

Rodrigodd opened the pull request:
Add a new module, similar to the waybar "sway/module" module. It shows the current mode under sway. For that, it needs to communicate with the Sway IPC. In this PR, the module is called "sway-mode" because "sway/mode" didn't work in my YAML config file (it would probably work if I surrounded it with quotes, but I didn't test).

This PR has two commits: the first one implements the module using a new sway connection. The second one refactors the sway `Client`, under `clients::compositor::sway`, to allow dynamically registering events, instead of only handling Workspace update ones. This second commit avoid creating a new socket for each module that needs to communicate with the Sway IPC, not sure if that is a problem to begin with.

Rodrigodd commented on src/clients/compositor/sway.rs:
```diff
@@ -155,3 +207,18 @@ impl From<WorkspaceEvent> for WorkspaceUpdate {
     }
 }
+
+fn sway_event_to_event_type(event: &Event) -> EventType {
```
This function could eventually be replaced by https://github.com/JayceFayne/swayipc-rs/pull/55

JakeStanger replied:
Cool, happy with that for now

JakeStanger commented on Cargo.toml:
```diff
@@ -73,9 +73,11 @@ volume = ["libpulse-binding"]
 workspaces = ["futures-lite"]
 "workspaces+all" = ["workspaces", "workspaces+sway", "workspaces+hyprland"]
-"workspaces+sway" = ["workspaces", "swayipc-async"]
+"workspaces+sway" = ["workspaces", "sway"]
```
I'd actually rather this *was* duplicated, as it keeps it explicit & if the sway module deps change in the future it avoids bringing in to the workspace.

```suggestion
"workspaces+sway" = ["workspaces", "swayipc-async"]
````

Rodrigodd replied:
Tried to do this, but `create_workspace_client` uses `sway::Client`, which is also is used by the `sway-mode` module. I can implement this change by carefully spreading `cfg(any(feature = "workspaces+sway", feature = "sway"))` in a couple of places, but I not sure if that is a win.

Rodrigodd replied:
Actually, currently `sway::Client` is under `clients::compositor`, which is gated by `workspaces`. I first need to move the `sway::Client` out of there, it does not need to be gated by `workspaces`.

JakeStanger commented on docs/modules/Sway-mode.md:
```diff
@@ -0,0 +1,74 @@
+Displays the current sway mode.
+
```
Should be obvious, but could you add a note block here to make it explicit that the module is for the Sway compositor only pls

JakeStanger commented on docs/modules/Sway-mode.md:
```diff
@@ -0,0 +1,74 @@
+
+## Styling
+
+| Selector | Description  |
+| -------- | ------------ |
+| `.label` | Label widget |
```
I think this section needs updating - assuming it was copy/pasted from the label docs? :P

JakeStanger commented on src/clients/compositor/sway.rs:
```diff
@@ -4,15 +4,31 @@ use color_eyre::{Report, Result};
+    task_state: Mutex<TaskState>,
+}
+
+impl std::fmt::Debug for Client {
```
What's the reason for the manual debug impl here?

Rodrigodd replied:
`WorkspaceClient` requires `Debug`, and `TaskState::listeners: Arc<Vec<(EventType, Box<SyncFn>)>>,` does not implement it, so I need to make a manual implementation, at least for `TaskState`. I could make the debug implementation list the events being listened, but I preferred to avoid locking `TaskState` when debug printing and just printing `<...>` instead. Let me know if you prefer otherwise.

JakeStanger replied:
Cool nope that makes sense & that's good with me. Figured it'd be something like that, just wnated to check.

JakeStanger commented on src/clients/compositor/sway.rs:
```diff
@@ -21,36 +37,59 @@ impl Client {
-            let workspace_tx = workspace_tx.clone();
+        // Only the task and self have a reference to listeners, and we just abort the task. This
+        // is the only reference to listeners, so we can safely unwrap.
+        let listeners_mut = Arc::get_mut(listeners).unwrap();
```
`.unwrap` should be [avoided](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap) in prod code. In part, it's because it is unclear that the situation *is* safe, without relying on a comment above. Generally if you're unwrapping because the code would have to get into an "impossible" state you should use `expect` instead.

In this case since you're returning a `Result` anyway, it would be better to return an error either with `?` or explicitly, and handle the situation at the call.

JakeStanger commented on src/modules/sway/mode.rs:
```diff
@@ -0,0 +1,94 @@
+                        let Event::Mode(mode) = event else {
+                            unreachable!()
+                        };
+                        let _ = await_sync(tx.send(ModuleUpdateEvent::Update(mode.clone())));
```
This can be simplified using `crate::try_send`:
```suggestion
                        try_send!(tx, ModuleUpdateEvent::Update(mode.clone()))
```

(I think I counted those brackets right...)

JakeStanger commented on src/modules/sway/mode.rs:
```diff
@@ -0,0 +1,94 @@
+                    EventType::Mode,
+                    Box::new(move |event| {
+                        trace!("event: {:?}", event);
+                        let Event::Mode(mode) = event else {
+                            unreachable!()
+                        };
```
I'm not a fan of requiring this when the type is (sorta) already specified in the call. I can't see how to avoid it nicely though. Any ideas?

Rodrigodd replied:
I created a `SwayIpcEvent` trait with a associated const `EVENT_TYPE: EventType` and a `fn from_event(&Event) -> &Self` method, implemented to all event types, and used that to fix this issue.

JakeStanger reviewed (changes requested):
Hey, thanks for this. It looks very good, just a few nits. 

Are you also able to provide some evidence of this working please? I can't get Sway to run at the moment :(

Rodrigodd commented:
I believe I have addressed all points. Here is a video of the sway mode working:


https://github.com/user-attachments/assets/20a3d3f2-2844-4956-a352-67d1ca1b6897

JakeStanger reviewed (commented):
Code all lgtm, cheers for making those changes & cheers for the vid.

One last thing before I approve - could you update `docs/sidebar.md` to add an entry for the docs page pls

Rodrigodd commented:
@JakeStanger done.

JakeStanger commented on src/clients/sway.rs:
```diff
@@ -0,0 +1,181 @@
+        self.add_listener_type(
+            T::EVENT_TYPE,
+            Box::new(move |event| {
+                let event = T::from_event(event).unwrap();
```
Oops no sorry, one more actually - can we remove this `unwrap` please - either `expect` if it should never fail, or safely handled otherwise

Rodrigodd replied:
Replace it by a `expect`. The type is enum variant is checked in `add_listener_type`, write after passing it to the closure.

Rodrigodd commented:
@JakeStanger all done. (just in case my last comment didn't trigger a notification)

JakeStanger commented:
Cheers!
